Смышляев Дмитрий

vafle228 opened the pull request:
@w1jtoo

Что на данный момент есть
1) Рабочий алгоритм генерации облаков
2) 2 варианта генераторов точек, для этих облаков
3) Генератор изображений и сохранялка для них

vafle228 committed:
Обновил структуру проекта

vafle228 committed:
Тест на инициализацию конструктора

vafle228 committed:
Вынес CloudLayouter как интерфейс

vafle228 committed:
Дискретная спираль фибоначи
- Добавил дискретную спираль
- Написал тесты на отрицательный шаг
- Написал тесты на штатную инициализацию
- Реализовал функционал тестов

vafle228 committed:
Генерация точек спирали
- Написал тесты для обхода по спирали
- Реализовал функционал обхода по спирали согласно тестам

vafle228 committed:
Забыл про Direction ...

vafle228 committed:
Размещение прямоугольников
- Отрефакторил код
- Добавил тест на добавление прямоугольника
- Реализовал логику согласно тесту

vafle228 committed:
Забыл про тест на пересечение
- Не бейте

vafle228 committed:
Добавил генерацию изображений

vafle228 committed:
Это был не Фибоначи ...
- Это был Архимед

vafle228 committed:
Новый генератор точек
- И новая директория под ниъ

vafle228 committed:
Еще игра в рефакторинг

vafle228 committed:
Спираль в полярных
- Ввел спираль в полярных координатах
- Отрефакторил код

vafle228 committed:
Негативные аргументы конструктора
- Написал тест на обработку негативных значений
- Добавил обработку негативных значений

vafle228 committed:
Генерация точек
- Добавил тест на генерацию точек
- Реализовал генерацию

vafle228 committed:
Сгенерированные картинки

vafle228 committed:
Убрал мусор

vafle228 committed:
Не тот мусор -_-

w1jtoo commented on cs/TagsCloudVisualization/SpiralLayouter/SpiralCloudLayouter.cs:
```diff
@@ -0,0 +1,39 @@
+            var rectangleUpperLeft = rectangleCenter - rectangleSize / 2;
+            var rect = new Rectangle(rectangleUpperLeft, rectangleSize);
+
+            if (!placedRectangles.Any(r => r.IntersectsWith(rect)))
```
Скорее `.All(!Intersects)`

w1jtoo commented on cs/TagsCloudVisualization/SpiralLayouter/SpiralCloudLayouter.cs:
```diff
@@ -0,0 +1,39 @@
+    private List<Rectangle> placedRectangles;
+    private IPointGenerator<Point> pointGenerator;
+    
+    public Point Center { get; private set; }
```
Зачем нужен setter?

w1jtoo commented on cs/TagsCloudVisualization/SpiralLayouter/PointGenerator/SquareArchimedesSpiral.cs:
```diff
@@ -0,0 +1,69 @@
+        return true;
+    }
+
+    public void Reset()
```
Смущает, что у тебя спираль на самом деле не просто спираль. Она ещё знает что-то про размещения.

Думаю, что это не так плохо, если ты честно будешь позиционировать этот класс как спираль и что-то для передвижения . И этот метод, тогда, должен делать смещение (а не Reset).

w1jtoo commented on cs/TagsCloudVisualizationTests/SpiralLayouter/PointGenerator/PolarArchimedesSpiralTest.cs:
```diff
@@ -0,0 +1,72 @@
+        }
+    }
+
+    [TestCaseSource(nameof(ThrowErrorOnNotPositiveNumberTestCases))]
```
Тут такое ощущение, что проще будет использовать `TestCase` обычный) Точка у тебя всегда (0, 0), а остальные параметры вполне себе можно положить в атрибут :)

w1jtoo commented on cs/TagsCloudVisualizationTests/SpiralLayouter/SpiralCloudLayouterTest.cs:
```diff
@@ -0,0 +1,66 @@
+        var random = new Random();
+        var layouter = new SpiralCloudLayouter(new Point(0, 0));
+        var rectangleSizes = new Size[100];
+        for (var i = 0; i < rectangleSizes.Length; i++)
```
Проверки на пересечение ну очень сложная) Давай проще

w1jtoo commented on cs/TagsCloudVisualizationTests/SpiralLayouter/PointGenerator/PolarArchimedesSpiralTest.cs:
```diff
@@ -0,0 +1,72 @@
+namespace TagsCloudVisualizationTests.SpiralLayouter.PointGenerator;
+
+[TestFixture]
+public class PolarArchimedesSpiralTest
```
Кажется, у тебя есть тест на какой-то один случай, когда спираль круглая/квадратная. 

Я думаю, что благодаря такому разделению ты можешь относительно просто проверить в этих тестах, что точки, которые генерирует спираль стремятся к окружности/квадрату.

w1jtoo reviewed (commented):
Подумай над требованиям по форме облака, как их можно отобразить в тестах

Если получится, можно отразить требование на плотность замощения прямоугольниками

vafle228 committed:
Переопределил интерфейс генераторов точек

vafle228 committed:
Переписал логику на новый интерфейс генератора

vafle228 committed:
Починил тесты

vafle228 committed:
Рефакторинг тестов

vafle228 committed:
Тесты на форму

vafle228 committed:
Тесты на плотность укладки

vafle228 committed:
Слегка поменял генерацию картинок
- BitmapGenerator теперь центрирует прямоугольники самостоятельно
- В пару к генератору появился BitmapSaver, который сохраняет и возвращает путь

vafle228 committed:
Добавил генерацию картинок для упавших случаев
